### PR TITLE
Tickets/sitcom 447 - Create a new script to run the warm up sequence for the hexapods

### PR DESCRIPTION
**Pull request by b1quint**

### CODE REVIEW

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from lsst.ts import salobj
+
+
+class WarmUpHexapod(salobj.BaseScript):
+    """Warm up Camera/M2 Hexapod after they have been idle for a long time. 
+
+    This sequence consists of moving the hexapod (camera or m2 hexapod) in 
+    one direction until it reaches its maximum position in both positive 
+    and negative directions. It moves in discrete steps to avoid
+    a FollowingError, which that happens when performing long movements.
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:34]: "which that happens" -> "which happens"

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    This sequence consists of moving the hexapod (camera or m2 hexapod) in 
+    one direction until it reaches its maximum position in both positive 
+    and negative directions. It moves in discrete steps to avoid
+    a FollowingError, which that happens when performing long movements. 
+    
+    This error became more evident in April 2022 and is reported in both 
+    [FRACAS-110] and [FRACAS-111]. 
+
+    You will probably run this script a couple of times with increasing
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:39]: "a couple of times" -> "a couple times"

Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:39]: Why not allow the Script to run with increasing step sizes? you could make `set_sizes` be an array and the default value consist of a sequence you found to be sufficient to work most of the time.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py (hunk 1 above)
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py]: Done.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py (hunk 2 above)
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py]: Done

Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py]: This is done now.

**Comment by b1quint** [comment]
I applied the comments and ran this script on TTS. It seems to be working fine.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+        # Assert step_size and sleep_time are compatible
+        if isinstance(self.config.step_size, (int, float)):
+            self.config.step_size = [self.config.step_size]
+
+        if isinstance(self.config.sleep_time, (int, float)):
+            self.config.sleep_time = [self.config.sleep_time]
+
+        if len(self.config.step_size) != len(self.config.sleep_time):
+            raise ValueError(
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:150]: For the error message consider dropping the "len"... You could do something like:

```
Expected same number of elements for step_size and sleep_time, got {} and {}, respectively.
```

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+                f"Expected `step_size` and `sleep_time` to have the same number of elements. "
+                f"Found len(step_size) ({len(config.step_size)}) different from "
+                f" len(sleep_time) = ({len(config.sleep_time)})."
+            )
+
+        # Relay the function used to execute the movement
+        if self.hexapod_name == "camera_hexapod":
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:157]: maybe do:

```
self.move_hexapod = self.mtcs.move_camera_hexapod if self.hexapod_name == "camera_hexapod" else self.mtcs.move_m2_hexapod
```

maybe use sal_index to compare though because then you don't need to rely on a hardcoded string here, e.g.;

```
self.move_hexapod = self.mtcs.move_camera_hexapod if self.hexapod_sal_index == SalIndex.CAMERA_HEXAPOD else self.mtcs.move_m2_hexapod
```

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        self.config = config
+
+        # Converts the string to salindex
+        self.hexapod_name = config.hexapod
+        self.sal_index = getattr(SalIndex, self.hexapod_name.upper())
+        self.hexapod = getattr(self.mtcs.rem, f"mthexapod_{self.sal_index}")
+        self.log.debug(f"Using the {self.hexapod_name} with sal index {self.sal_index}")
+
+        # Assert step_size and sleep_time are compatible
+        if isinstance(self.config.step_size, (int, float)):
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:143]: It might be more reliable to check it `step_size` is iterable, e.g.;

```
from collections.abc import Iterable

if not isinstance(self.config.step_size, Iterable):
    ...
```

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        metadata.duration = sum(
+            [
+                time * self.get_number_of_steps(step)
+                for time, step in zip(self.config.sleep_time, self.config.step_size)
+            ]
+        )
+
+    def get_number_of_steps(self, step_size):
+        """
+        Returns the total number of steps depending on the step size considering
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py]: Start this phrase in the line above.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                raise
+        await self.checkpoint("Done")
+
+    async def warm_up(self):
+        for step, sleep_time in zip(self.config.step_size, self.config.sleep_time):
+            await self.single_loop(step, sleep_time)
+
+    async def move_stepwise(self, start, stop, step, sleep_time):
+        """
+        Moves the hexapod from `start` to `stop` in `steps`.
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:205]: Start this phrase in the line above.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+            await self.checkpoint(checkpoint)
+            try:
+                await activity()
+            except Exception:
+                self.log.exception(f"Error running checkpoint: {checkpoint}")
+                raise
+        await self.checkpoint("Done")
+
+    async def warm_up(self):
+        for step, sleep_time in zip(self.config.step_size, self.config.sleep_time):
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:200]: Please, add a docstring.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+            all_positions = {axes[i]: pos.position[i] for i in range(len(axes))}
+            all_positions[self.config.axis] = new_pos
+
+            await self.move_hexapod(**all_positions)
+            await asyncio.sleep(sleep_time)
+
+    async def single_loop(self, step, sleep_time):
+        """
+        Do a full loop moving from the current position to the positive limit position,
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py]: Start this phrase in the line above...

File: tests/maintel/test_warmup_hexapod.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+index_gen = utils.index_generator()
+
+logger = logging.getLogger(__name__)
+logger.propagate = True
+
+
+class TestCameraHexapodWarmUp(
+    standardscripts.BaseScriptTestCase, unittest.IsolatedAsyncioTestCase
+):
+    def setUp(self):
```
Comment (tribeiro) [inline tests/maintel/test_warmup_hexapod.py:43]: Looks like you don't need this, can you just remove it?

File: tests/maintel/test_warmup_hexapod.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        pass
+
+    async def basic_make_script(self, index):
+        logger.debug("Starting basic_make_script")
+        self.script = WarmUpHexapod(index=index)
+
+        # self.script.checkpoints_activities = [
```
Comment (tribeiro) [inline tests/maintel/test_warmup_hexapod.py:50]: Please, remove commented code.

File: tests/maintel/test_warmup_hexapod.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+from lsst.ts import utils
+from lsst.ts import standardscripts
+from lsst.ts.externalscripts.maintel import WarmUpHexapod
+from lsst.ts.idl.enums import Script
+
+np.random.seed(42)
+
+index_gen = utils.index_generator()
+
+logger = logging.getLogger(__name__)
```
Comment (tribeiro) [inline tests/maintel/test_warmup_hexapod.py:36]: Maybe instead of creating a logger here, add log to the test case class.. you can do it in the `setUp` method...

File: tests/maintel/test_warmup_hexapod.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            assert self.script.config.step_size == [
+                step_size
+            ]  # int/float are converted to lists
+            assert self.script.config.sleep_time == [
+                sleep_time
+            ]  # int/float are converted to lists
+            assert self.script.config.max_position == max_position
+
+            assert self.script.hexapod_name == hexapod
+            assert self.script.sal_index == 1  # camera_hexapod
```
Comment (tribeiro) [inline tests/maintel/test_warmup_hexapod.py:89]: Consider checking agains the enumerations.. This way you don't have to update the test if we change the value.

File: tests/maintel/test_warmup_hexapod.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                position = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+
+            self.script.move_hexapod = unittest.mock.AsyncMock()
+            self.script.hexapod.tel_application.aget = unittest.mock.AsyncMock(
+                return_value=(MockPosition)
+            )
+
+            # Run the script
+            await self.run_script()
+            assert self.script.state.state == Script.ScriptState.DONE
```
Comment (tribeiro) [inline tests/maintel/test_warmup_hexapod.py:125]: Can you add a `test_executable` test? Check this one https://github.com/lsst-ts/ts_standardscripts/blob/e9b64618b318c9c0dfdba950614c469949af630e/tests/test_run_command.py#L131-L134

**Review by tribeiro — COMMENTED** [review]
Here some additional comments.. Thanks for the updates it is looking pretty good...

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+        self.mtcs = (
+            MTCS(domain=self.domain, log=self.log)
+            if remotes
+            else MTCS(
+                domain=self.domain, log=self.log, intended_usage=MTCSUsages.DryTest
+            )
+        )
+
+        # The estimated duration deppends on the configuration
+        self.activity_estimated_duration = None
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:69]: I am not using this attribute. So I will remove it.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+              step_size:
+                description: >-
+                  The discrete step size in microns in which the hexapod will move. 
+                  This can be a number or an array of numbers.
+                anyOf:
+                  - type: number
+                  - type: array
+                default: 250
+              sleep_time:
+                description: The sleep time in seconds between movements movements.
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:103]: Solved.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+                anyOf:
+                  - type: number
+                  - type: array
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:106]: The problem here is that we have a allow different sleep times for different step sizes. For example, the sleep time for 100um movements can be 1s while the sleep time for 1000 um should be around 5s. So, yes, we need it to be an array or a number. I added more documentation saying that the number of elements of the `sleep_time` should match the number of elements of `step_size`.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+                default: 1
+              max_position:
+                description: Maximum position (absolute value) for the movements.
+                type: number
+                default: 13000
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:111]: Done.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+            f"  step_size: {config.step_size}\n"
+            f"  sleep_time: {config.sleep_time}\n"
+            f"  max_position: {config.max_position}\n"
+        )
+
+        self.config = config
+
+        # Converts the string to salindex
+        self.hexapod_name = config.hexapod
+        self.sal_index = getattr(SalIndex, self.hexapod_name.upper())
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:138]: Done.

File: tests/maintel/test_warmup_hexapod.py (hunk 10 above)
Comment (b1quint) [reply tests/maintel/test_warmup_hexapod.py:43]: Done.

File: tests/maintel/test_warmup_hexapod.py (hunk 11 above)
Comment (b1quint) [reply tests/maintel/test_warmup_hexapod.py:50]: Done.

File: tests/maintel/test_warmup_hexapod.py (hunk 10 above)
Comment (b1quint) [reply tests/maintel/test_warmup_hexapod.py:43]: Actually, added the code to setup the logger here.

File: tests/maintel/test_warmup_hexapod.py (hunk 14 above)
Comment (b1quint) [reply tests/maintel/test_warmup_hexapod.py:119]: Done.

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+            $schema: http://json-schema/draft-07/schema#
+            $id: https://github.com/lsst-ts/ts_standardscripts/maintel/WarmUpHexapod.yaml
+            title: WarmUpHexapod v1
+            description: Configuration for WarmUpHexapod.
+            type: object
+            properties:
+              hexapod:
+                description: Which hexapod will move? (camera_hexapod/m2_hexapod)
+                type: string
+                enum: ["camera_hexapod", "m2_hexapod"]
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:85]: Done.

File: tests/maintel/test_warmup_hexapod.py (hunk 10 above)
Comment (tribeiro) [reply tests/maintel/test_warmup_hexapod.py:41]: Nice, Picky comment but we usually name this `log` not `logger`..

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py (hunk 3 above)
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:157]: done

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+            self.config.step_size = [self.config.step_size]
+
+        if isinstance(self.config.sleep_time, (int, float)):
+            self.config.sleep_time = [self.config.sleep_time]
+
+        if len(self.config.step_size) != len(self.config.sleep_time):
+            raise ValueError(
+                f"Expected `step_size` and `sleep_time` to have the same number of elements. "
+                f"Found len(step_size) ({len(config.step_size)}) different from "
+                f" len(sleep_time) = ({len(config.sleep_time)})."
```
[the remark this answers is not in the document]
Comment (b1quint) [reply python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:160]: this is solved with the suggestion above.

File: tests/maintel/test_warmup_hexapod.py [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+            assert self.script.hexapod_sal_index == getattr(SalIndex, hexapod.upper())
+
+    async def test_run(self):
+
+        # Start the test itself
+        async with self.make_script():
+
+            # Try configure with minimum set of parameters declared
+            # Note that all are scalars and should be converted to arrays
+            hexapod = "camera_hexapod"
```
Comment (tribeiro) [inline tests/maintel/test_warmup_hexapod.py:92]: This should be `hexapod = "camera"` only now

File: python/lsst/ts/externalscripts/maintel/warmup_hexapod.py [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+
+        # Converts the string to salindex
+        self.hexapod_name = f"{config.hexapod}_hexapod"
+        self.hexapod_sal_index = getattr(SalIndex, self.hexapod_name.upper())
+        self.hexapod = getattr(self.mtcs.rem, f"mthexapod_{self.hexapod_sal_index}")
+        self.log.debug(
+            f"Using the {self.hexapod_name} with sal index {self.hexapod_sal_index}"
+        )
+
+        # Assert step_size and sleep_time are compatible
```
Comment (tribeiro) [inline python/lsst/ts/externalscripts/maintel/warmup_hexapod.py:153]: "Assert" usually means we might expect an `AssertionError`, which is not really happening here. 

Maybe reword to like:

> Make sure step_size and sleep_time are arrays.

**Review by tribeiro — COMMENTED** [review]
A couple more comments

**Review by tribeiro — APPROVED** [review]
Looks good.. please,  make sure you take a look at my latest comments and that the CI job is passing. 

Thanks for all the effort!
